### DISCUSSION TITLE
Pass in a hash for a value

### DISCUSSION THREAD

**Discussion by adrianthedev** [josefarias/hotwire_combobox#221] [Q&A]
Hey 👋

Thanks for the awesome project Jose!

Quick question:
Is it possible to pass in a hash for a value to hw_combobox?

Our peculiar use-case at https://github.com/avo-hq/avo is that we'd like to create a combobox field.
The API looks kind of like this:

```ruby
field :location,
  as: :combobox,
  query: -> {
    # return your array of display & value hashes
    City.search(params[:q])
      .map do |city|
        {
          display: city.name,
          value: city.id
        }
      end
  },
  find_value: -> {
    # find your record and tell us what's the display value
    city = City.find(value)

    {
      display: city.name,
      value: city.id
    }
  }
```

I foresee two use-cases:

1. When the user searches for a string and the stored value will be that string.
This should be fairly easy to support with the current customizations

2. The user searches for a record, stores the ID (what HWC uses for `value`) but wants to display some kind of label for that record (what HWC uses for `display`).

Ideally, we'd like to be able to use the combobox like this.

```ruby
# @field.id = "name"
# model.name = 30
@form.combobox @field.id, # "name"
  helpers.combobox_results_path(resource_class: @resource.class, field_id: @field.id),
  id: "#{@field.id}_id",
  value: {display: "Barcelona", value: 30}
```

The `value` (naming can be changed) would return a hash or some other pattern that HWC would use, to enable the user to give whichever hash value they want.

Is this something that can be done with HWC and maybe I missed it in the docs?

PS: I know about how HWC is searching for the association in the form model based on the `_id` suffix, but our users don't always have that association in the form model.

PPS: I read https://github.com/josefarias/hotwire_combobox/discussions/167 and it's not that scenario.

PPPS: In case you're wondering why the model might not have the association object attached for the regular way of how HWC works, we sometimes use these fields on actions and other entities which are not AR objects.

PPPPS: [This](https://github.com/avo-hq/avo/pull/3381) is the PR that adds HWC to Avo.

Thanks 🙏

**Comment by adrianthedev** [comment]
@josefarias could you please weigh in here?

TL;DR; is there a way to send a hash/object as the value to the combobox field?

**Comment by josefarias** [comment]
Hey @adrianthedev, sorry I had this on the backburner as I needed to learn more about Avo and what `fields` are in Avo's context. I think I understand what you'd like to do now.

Answering the question directly, I think the most reliable way to pass hashes as values would be to serialize them as strings. Maybe with https://github.com/hopsoft/universalid? (Or simply as JSON) I haven't used that gem but it looks very good. Could also serialize any kind of struct if hashes don't work.

That said – I'm not sure this'd be an implementation I recommend. I must say I don't yet understand the full implications of this feature. But it strikes me as smelly that you seem to want to serialize the display of objects you have access to, when that display should already be available via `#to_combobox_display`.

Or, if what you're aiming to do is persist the query (and not the `display` or `value`) you should be able to obtain that from the custom events fired by the combobox and write them to a separate hidden field, or inject them into a payload upon submission. Sorry the events are not documented yet, but they're visible in the gem's source under `events.js`

**Comment by adrianthedev** [comment]
Yeah, it seems smelly, but we do have use-cases for it.
One use case that we're trying to use it right now is to fetch the records from an HTTP endpoint and feed them to the combobox results.
They are not AR objects, but plain Hashes.

I made a video to show you the problem. Maybe you see other ideas on how to fix it.
https://www.loom.com/share/34aa420250bf4296a0c9b48e8f908369?sid=81188d91-3367-4016-9549-868327d55b60

Thanks!

**Reply by josefarias** [reply]
Ah thanks for the video @adrianthedev, that's very helpful!

So it seems in this example you're using the combobox to determine a title based on a collection of products. We should be able to sort that out!

I'm sorry I don't know enough about Avo internals to suss out the implementation. But I believe in your video you're providing either an AR relation or an array of ActiveRecords as options, correct? What's happening is, when provided with such options, HotwireCombobox assumes you'll want the `value` to be the result of calling `#id` on each option. That's the most common use case and thus the default, but it's configurable.

I think the way to go is passing an array of hashes as the options (instead of an AR relation or an array of ARs). Something like: `Product.all.map { { display: _1.name, value: _1.name } }` (of course that can be paginated and provided async, this is just an example). That should make it so the `value` when choosing an option is the result of `#name`, and not `#id`. And so the name gets sent as a string when submitting the form. There's more information about what you can pass as options in the "basic combobox" section of the docs at [hotwirecombobox.com](https://hotwirecombobox.com/). Looks like this:

```rb
@options = Product.all.map { { display: _1.name, value: _1.name } }
<%= f.combobox :name, @options, ... %>
```

If you're unable to manipulate the options before feeding them to the combobox, and so you absolutely need to pass a relation as the options — there's another solution. But I'd consider this slightly more advanced usage and I have no plans to document it. That'd be passing the name of a method to the `combobox_options` helper. So something like this: `f.combobox :field_name, combobox_options(@relation, value: @method_name)`. In the example above `@method_name` would be `:name` and it'd be called by the gem [here](https://github.com/josefarias/combobox/blob/0fa9a7f3fbacd36a2d8616933c44a1f2d07b047f/app/presenters/hotwire_combobox/listbox/item.rb#L73) — I'm guessing that'd be provided by the user or something. But if you know the name of the method offhand, and you're able to manipulate the options, I'd go with the first option. Otherwise I'd suggest getting familiar with the `combobox_options` helper because you'll be opting out of some syntax sugar.

Hope that helps! Let me know if not, we can keep talking through it.

**Comment by Paul-Bob** [comment]
Hi @josefarias

We tried to extract the issue from the Avo context and I've created an reproduction repository where it's reproducible outside the Avo context.

We used an HTTP endpoint just to show an example on how it can be outside the scope of ActiveRecord associations.

This is the [repo](https://github.com/avo-hq/hotwire_combobox_value)  which has all the steps for the setup and reproduction.  (Please let me know if something doesn't work as expected)

The end goal would be to be able to pass an hash (or any other object that responds to `value` and `display`) to the combobox form helper, something like this:

```ruby
<%= form.combobox :status, hotwire_combobox_options_path, value: { value: 1, display: "One" } %>
```

We also have a minimal POC (on hotwire_combobox) which would make this work but we didn't push it yet, let me know if you want us to push it and discuss on the approach.

Thank you!

**Reply by josefarias** [reply]
Hey @Paul-Bob! Thanks for the repo, I can check it out later today. In the meantime, could you confirm whether my suggestions from https://github.com/josefarias/hotwire_combobox/discussions/221#discussioncomment-11360503 help at all? I suspect something along those lines may be all you need.

**Reply by adrianthedev** [reply]
Hey Jose. Paul and I are working together and I asked him to post that repro we did together.

I don't think the suggestions from your previous answer apply. I think the repro will show you the use-case I was talking about.
If that doesn't relay the message, let us know, and will talk more about it.

**Comment by josefarias** [comment]
@Paul-Bob, @adrianthedev I made some changes to your repo and I believe I've got it working as expected. Here's a quick video:

https://github.com/user-attachments/assets/47a6137c-a317-4bb0-b2be-4799df44405c

Here are the changes. Note that the changes to `application.js` and `importmap.rb` are necessary, but for a different reason. Without them, you'd be loading the combobox code twice, resulting in some unrelated wonkiness. For the issue raised in this discussion, you'll want to look at the changes in `posts_controller.rb` and `_form.html.erb`.

I do not believe you need to pass a hash as a value for this to work. I think some variation of https://github.com/josefarias/hotwire_combobox/discussions/221#discussioncomment-11360503 should be enough.

```diff
diff --git a/app/controllers/posts_controller.rb b/app/controllers/posts_controller.rb
index b793e4b..17af328 100644
--- a/app/controllers/posts_controller.rb
+++ b/app/controllers/posts_controller.rb
@@ -16,7 +16,7 @@ def hotwire_combobox_options
 
     options = data["products"].map do |item|
       {
-        value: item["id"],
+        value: item["title"],
         display: item["title"]
       }
     end
diff --git a/app/javascript/application.js b/app/javascript/application.js
index 5307264..1a6de45 100644
--- a/app/javascript/application.js
+++ b/app/javascript/application.js
@@ -2,11 +2,6 @@
 import "@hotwired/turbo-rails"
 import "controllers"
 import { Application } from "@hotwired/stimulus"
-import HwComboboxController from '@josefarias/hotwire_combobox'
-
 const application = Application.start()
 
-// Add the following two lines:
-application.register("hw-combobox", HwComboboxController)
-
 export { application }
diff --git a/app/views/posts/_form.html.erb b/app/views/posts/_form.html.erb
index 847e245..ca5c0bc 100644
--- a/app/views/posts/_form.html.erb
+++ b/app/views/posts/_form.html.erb
@@ -18,26 +18,7 @@
 
   <div>
     <%= form.label :status, style: "display: block" %>
-    <%
-      value = if post.status.present?
-        # Ideally we should be able to pass the value as a hash or any object that responds to `value` & `display`, or any other API that you think it's good
-        url = URI("https://dummyjson.com/products/#{post.status}")
-
-        # Make the HTTP request
-        response = Net::HTTP.get(url)
-
-        # Parse JSON response
-        data = JSON.parse(response)
-      
-        {
-          value: data["id"],
-          display: data["title"]
-        }
-      else
-        nil
-      end
-    %>
-    <%= form.combobox :status, hotwire_combobox_options_path, value: value %>
+    <%= form.combobox :status, hotwire_combobox_options_path, value: post.status %>
   </div>
 
   <div>
diff --git a/config/importmap.rb b/config/importmap.rb
index ea07134..7fdea4a 100644
--- a/config/importmap.rb
+++ b/config/importmap.rb
@@ -5,4 +5,3 @@
 pin "@hotwired/stimulus", to: "@hotwired--stimulus.js" # @3.2.2
 pin "@hotwired/stimulus-loading", to: "stimulus-loading.js"
 pin_all_from "app/javascript/controllers", under: "controllers"
-pin "@josefarias/hotwire_combobox", to: "@josefarias--hotwire_combobox.js" # @0.3.2
```

**Reply by Paul-Bob** [reply]
We see what you've done here, and this aligns perfectly with a use case where the label is directly stored in the database. 

However, our limitation is that we have the ID stored in the DB but we want to show the label in the input while keep the ID as the input value when the form is submitted.

**Reply by josefarias** [reply]
@Paul-Bob I think you should be able to get by by denormalizing the data into a hidden field. Maybe like this:

```diff
diff --git a/app/javascript/controllers/denormalize_controller.js b/app/javascript/controllers/denormalize_controller.js
new file mode 100644
index 0000000..56fec50
--- /dev/null
+++ b/app/javascript/controllers/denormalize_controller.js
@@ -0,0 +1,9 @@
+import { Controller } from "@hotwired/stimulus"
+
+export default class extends Controller {
+  static targets = [ "input" ]
+
+  update(event) {
+    this.inputTarget.value = event.detail.display
+  }
+}
diff --git a/app/views/posts/_form.html.erb b/app/views/posts/_form.html.erb
index 847e245..b17a202 100644
--- a/app/views/posts/_form.html.erb
+++ b/app/views/posts/_form.html.erb
@@ -16,28 +16,10 @@
     <%= form.text_field :name %>
   </div>
 
-  <div>
+  <div data-controller="denormalize">
     <%= form.label :status, style: "display: block" %>
-    <%
-      value = if post.status.present?
-        # Ideally we should be able to pass the value as a hash or any object that responds to `value` & `display`, or any other API that you think it's good
-        url = URI("https://dummyjson.com/products/#{post.status}")
-
-        # Make the HTTP request
-        response = Net::HTTP.get(url)
-
-        # Parse JSON response
-        data = JSON.parse(response)
-      
-        {
-          value: data["id"],
-          display: data["title"]
-        }
-      else
-        nil
-      end
-    %>
-    <%= form.combobox :status, hotwire_combobox_options_path, value: value %>
+    <%= form.hidden_field :display, data: { denormalize_target: "input" } %>
+    <%= form.combobox :status, hotwire_combobox_options_path, value: post.status, data: { action: "hw-combobox:selection->denormalize#update" } %>
   </div>
 
   <div>
```

Of course you're free to persist as a stringified hash 😄 via `to_json` or the gem I shared originally, or any other method. You'd just have to deserialize as needed, too.

I think this conversation is more about the domain model than the combobox itself though so I'm afraid I can't go much deeper into it without getting familiar with the domain first. But if you absolutely have to denormalize, I think a hidden field might be less involved than serializing/deserializing hashes.

**Reply by adrianthedev** [reply]
Hey Jose,

Sorry, went to work on a different thing and we didn't get back to you.

I guess, the domain is somehow ismilar to the `select` input domain.
At least that's how we look at it. Possibly others too.

We'd basically want to substitute a regular `select` field with HWC. The reasoning is that we'd like to gain the async behavior that the regular `select` doesn't have.
In that mental model, we're used to render a list of options and the actual value of the field and it knows how to display the label of that option instead of the value.

That's basically the same thing we'd like to do, with the caveat that because of the `async` behavior HWC might not have the label for that value, and that's perfectly understandable.
This is why we'd like to augment that behavior by giving it the value and also the label to HWC for it to display to the user.

**Reply by adrianthedev** [reply]
BTW, We're open to a call as well, if that makes more sense to you.

**Reply by josefarias** [reply]
Sorry @adrianthedev I'm just not seeing the problem 😅 that should be achievable with the combobox as it works today.

**Reply by josefarias** [reply]
@adrianthedev here's an idea. Could you maybe build an example app with the standard select flow you're aiming to replace? And I could suggest how to implement that with HWCB. Or maybe I'll finally see that it's not possible and understand what you're missing here.

**Reply by adrianthedev** [reply]
Got it!
Tomorrow is release day and we're a bit caught in that, but we will do that comparison and I'll com up with a new video to show it better.

**Reply by adrianthedev** [reply]
Thanks Jose 🙏

**Reply by josefarias** [reply]
Np! Very open to the possibility I'm missing something here I think the example app will help.

I wouldn't want you to spend much time on it. Just a working, faithful, example where a plain select works as intended should be enough. I can build a HWCB equivalent from that. 

I think where the last reproduction app lost me was that you built it with HWCB so I couldn't see what the expected behavior was exactly, I could only interpret it from our conversations. But seeing it work as expected should help with that.
